---
title: "OSCA - SingleCellExperiment"
author: "Chun-Jie Liu"
date: "2022-05-03"
---

> Seurat is very high integrated package. Actually, the basic workflow of scRNA-seq analysis is:

1. We compute quality control metrics to remove low-quality cells that would interfere with downstream analyses. These cells may have been damaged during processing or may not have been fully captured by the sequencing protocol. Common metrics includes the total counts per cell, the proportion of spike-in or mitochondrial reads and the number of detected features.
2. We convert the counts into normalized expression values to eliminate cell-specific biases (e.g., in capture efficiency). This allows us to perform explicit comparisons across cells in downstream steps like clustering. We also apply a transformation, typically log, to adjust for the mean-variance relationship.
3. We perform feature selection to pick a subset of interesting features for downstream analysis. This is done by modelling the variance across cells for each gene and retaining genes that are highly variable. The aim is to reduce computational overhead and noise from uninteresting genes.
4. We apply dimensionality reduction to compact the data and further reduce noise. Principal components analysis is typically used to obtain an initial low-rank representation for more computational work, followed by more aggressive methods like tstochastic neighbor embedding for visualization purposes.
5. We cluster cells into groups according to similarities in their (normalized) expression profiles. This aims to obtain groupings that serve as empirical proxies for distinct biological states. We typically interpret these groupings by identifying differentially expressed marker genes between clusters.

## Test code


Example code from [here](http://bioconductor.org/books/3.15/OSCA.intro/getting-scrna-seq-datasets.html#overview-1)

```{r}
library(scRNAseq)
```
```{r}
sce.zeisel <- ZeiselBrainData()
```

```{r}
library(scater)
```

```{r}
library(BiocFileCache)
bfc <- BiocFileCache(ask=FALSE)
url <- file.path("ftp://ftp.ncbi.nlm.nih.gov/geo/series",
    "GSE85nnn/GSE85241/suppl",
    "GSE85241%5Fcellsystems%5Fdataset%5F4donors%5Fupdated%2Ecsv%2Egz")
muraro.fname <- bfcrpath(bfc, url)
local.name <- URLdecode(basename(url))
unlink(local.name)
if (.Platform$OS.type=="windows") {
    file.copy(muraro.fname, local.name)
} else {
    file.symlink(muraro.fname, local.name)
}
```
```{r}
mat <- as.matrix(read.delim("GSE85241_cellsystems_dataset_4donors_updated.csv.gz"))
```
```{r}
dim(mat)
mat[1:4,1:4]
class(mat)
object.size(mat)
```

```{r}
library(scuttle)
sparse.mat <- readSparseCounts(local.name)
```
```{r}
dim(sparse.mat)
sparse.mat[1:4,1:4]
class(sparse.mat)
object.size(sparse.mat)
```

```{r}
library(DropletUtils)
sce <- read10xCounts("/home/liuc9/tmp/singlecell/filtered_gene_bc_matrices/hg19")
```

```{r}
sce
```
## `SingleCellExperiment` class
![sce](http://bioconductor.org/books/3.15/OSCA.intro/images/SingleCellExperiment.png)

```{r}
sce@colData
sce@metadata

```

```{r}
barcodes <- readr::read_tsv(file = "/home/liuc9/tmp/singlecell/filtered_gene_bc_matrices/hg19/barcodes.tsv", col_names = FALSE)
features <- readr::read_tsv(file = "/home/liuc9/tmp/singlecell/filtered_gene_bc_matrices/hg19/genes.tsv", col_names = FALSE)
mtx <- Matrix::readMM(file = "/home/liuc9/tmp/singlecell/filtered_gene_bc_matrices/hg19/matrix.mtx")
rownames(mtx) <- features$X1
colnames(mtx) <- barcodes$X1
```

```{r}
sce <- SingleCellExperiment(
  assays = list(counts = mtx)
)
```

```{r}
sce <- scuttle::logNormCounts(sce)
```
```{r}
counts_100 <- counts(sce) + 100
assay(sce, "counts_100") <- counts_100
# assay(sce)
sce
```
```{r}
assays(sce) <- assays(sce)[1:2]
```
```{r}
sce
```
```{r}
assayNames(sce)
```
## Dimensionality reduction
```{r}
library(DropletUtils)
sce <- read10xCounts("/home/liuc9/tmp/singlecell/filtered_gene_bc_matrices/hg19")
```
```{r}
sce %>%
  scater::logNormCounts() %>%
  scater::runPCA() ->
  sce
```

```{r}
sce
```
```{r}
reducedDim(sce, "PCA") %>% dim()
```

```{r}
sce <- scater::runTSNE(sce, perplexity = 0.1)
```

```{r}
head(reducedDim(sce, "TSNE"))
```

```{r}
sce
```
```{r}
reducedDim(sce, "TSNE") %>% head()
```
```{r}
reducedDims(sce)
```

```{r}
u <- uwot::umap(t(logcounts(sce)), n_neighbors = 2)
```
```{r}
sce <- scater::runUMAP(sce)
```
```{r}
reducedDims(sce)
```

## size factor

```{r}
sce <- scran::computeSumFactors(sce)
```
```{r}
sce
```
```{r}
colLabels(sce) <-  scran::clusterCells(sce, use.dimred = "PCA")
```

```{r}
colLabels(sce) %>% table
```

## Analysis overview

![Outline](http://bioconductor.org/books/3.15/OSCA.intro/images/workflow.png)

```{r}
sce <- scRNAseq::MacoskoRetinaData()
```
```{r}
sce
```

```{r}
is.mito <- grepl("^MT-", rownames(sce))
qcstats <- scater::perCellQCMetrics(sce, subset = list(Mito = is.mito))
filtered <- scater::quickPerCellQC(qcstats, percent_subsets="subsets_Mito_percent")
```

```{r}
sce <- sce[, !filtered$discard]
```

```{r}
sce
```

```{r}
sce <- scater::logNormCounts(sce)
```
```{r}
sce
sce@colData %>% head()
```

```{r}
dec <- scran::modelGeneVar(sce)
hvg <- scran::getTopHVGs(dec, prop = 0.1)
```
```{r}
dec
```

```{r}
set.seed(1234)
```
```{r}
sce <- scater::runPCA(sce, ncomponents=25, subset_row=hvg)
```

```{r}
sce
```

```{r}
colLabels(sce) <- scran::clusterCells(
  x = sce,
  use.dimred = "PCA",
  BLUSPARAM = bluster::NNGraphParam(
    cluster.fun = "louvain"
  )
)
```
```{r}
sce$label %>% table()
```


```{r}
sce <- scater::runUMAP(sce, dimred="PCA")
```

```{r}
sce
```
```{r}
reducedDim(sce, "UMAP") %>% head()
```

```{r}
scater::plotUMAP(sce, colour_by = "label")
```

```{r}
# Marker detection.
markers <- scran::findMarkers(sce, test.type="wilcox", direction="up", lfc=1)
```
```{r}
dim(sce)
sce$label %>% table()
markers$`1` %>% dim()
markers$`1`
```

## Multiple batches

```{r}
sce <- scRNAseq::SegerstolpePancreasData()
```

```{r}
qcstats <- scater::perCellQCMetrics(sce)
filtered <- scater::quickPerCellQC(qcstats, percent_subsets="altexps_ERCC_percent")
sce <- sce[, !filtered$discard]
```

```{r}
sce <- scater::logNormCounts(sce)
```

```{r}
dec <- scran::modelGeneVar(sce, block=sce$individual)
hvg <- scran::getTopHVGs(dec, prop = 0.1)
```
```{r}
dec %>% dim()
```

```{r}
set.seed(1234)
```
```{r}
sce <- batchelor::correctExperiments(
  sce,
  batch=sce$individual,
  subset.row=hvg,
  correct.all=TRUE
)
```

```{r}
colLabels(sce) <- scran::clusterCells(sce, use.dimred='corrected')

```


```{r}
sce <- scater::runUMAP(sce, dimred="corrected")
```

```{r}
sce
```

```{r}
gridExtra::grid.arrange(
    scater::plotUMAP(sce, colour_by="label"),
    scater::plotUMAP(sce, colour_by="individual"),
    ncol=2
)
```

```{r}
# Marker detection, blocking on the individual of origin.
markers <- scran::findMarkers(sce, test.type="wilcox", direction="up", lfc=1)
```

